lightning/src/routing/router.rs

```diff
@@ -349,63 +394,170 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 						// bit lazy here. In the future, we should pull them out via our
 						// ChannelManager, but there's no reason to waste the space until we
 						// need them.
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
+						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, None::<u64>, features.to_context(), 0);
 					}
 				}
 				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
 				// really sucks, cause we're gonna need that eventually.
-				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
+				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, None::<u64>, ChannelFeatures::empty(), 0);
 			}
 		}
 	}
 
-	while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, .. }) = targets.pop() {
-		if pubkey == *our_node_id {
-			let mut res = vec!(dist.remove(&our_node_id).unwrap().3);
-			loop {
-				if let Some(&(_, ref features)) = first_hop_targets.get(&res.last().unwrap().pubkey) {
-					res.last_mut().unwrap().node_features = features.to_context();
-				} else if let Some(node) = network.get_nodes().get(&res.last().unwrap().pubkey) {
-					if let Some(node_info) = node.announcement_info.as_ref() {
-						res.last_mut().unwrap().node_features = node_info.features.clone();
+	let mut payment_paths = Vec::new();
+	let mut tries_left = 10;
+	let mut found_amount_msat = 0;
+	// For every MPP start with the same pre-filled data.
+	let prefilled_targets = targets.clone();
```

Comment:
Does this mean we'll re-use last-hop and first-hop channels that we already "used" fully via `used_channsl_with_amounts` when calculating each path?

Reply:
Right, first hop and last hop channels are sort of pain in the ass. What are the expectations we're setting? Consider the following scenarios:
- we got 3 first_hops, and managed to route with 2 paths. Must both found paths use first_hops?
- we got 3 first_hops, and managed to route with 4 paths. Must all 4 found paths use first_hops?

Same question for the last_hops i guess.

Reply:
Right, I admit I don't know the answer here, though it seems strange to mark every channel except first+last as "we used X of this channel" - as written I could see the algorithm picking 5 paths each that use the same first hop, only to find out that the first hop didn't have enough balance and failing, while we have a (more expensive) first hop that we could have used.

Can you not move the add-first-last-hops stuff into the loop and the check the HashMap in the relevant macro?

Reply:
I understood the problem and tried to address it in the latest commit.

I'm not sure I understood your suggestion though, I think it has a typo or something. Anyway, perhaps take a look at my solution, and then if that doesn't work, please rephrase your suggestion :)